### PR TITLE
Enable `CA1067`: Override `Object.Equals(object)` when implementing `IEquatable<T>`

### PR DESCRIPTION
**Pull request by xtqqczze**
https://docs.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/ca1067

### CODE REVIEW

**Comment by iSazonov** [comment]
> Should an override be provided for `GetHashCode()` too?

It is not clear if we force the rule why does it not complain?

**Comment by xtqqczze** [comment]
> > Should an override be provided for `GetHashCode()` too?
> 
> It is not clear if we force the rule why does it not complain?

We might expect to see [compiler warning (level 3) CS0659](https://docs.microsoft.com/dotnet/csharp/misc/cs0659) if the override for `GetHashCode` is missing. It appears we have [level 5 warnings](https://docs.microsoft.com/dotnet/csharp/language-reference/compiler-options/warn-compiler-option) enabled.

**Comment by xtqqczze** [comment]
> > Should an override be provided for `GetHashCode()` too?
> 
> It is not clear if we force the rule why does it not complain?

The method already exists: 😛
`public override int GetHashCode()`

**Review by iSazonov — APPROVED** [review]

**Comment by iSazonov** [comment]
I am not sure it is a breaking change.

**Comment by xtqqczze** [comment]
> I am not sure it is a breaking change.

According to .NET compatibility guidelines, probably not:
> ALLOWED: Adding or removing an override

https://docs.microsoft.com/dotnet/core/compatibility/#members

**Comment by iSazonov** [comment]
Ask PowerShell Committee about adding new public API.

**Comment by SteveL-MSFT** [comment]
@PowerShell/powershell-committee reviewed this, we are ok taking the new public API, but a test needs to be added and we believe the current implementation may not be correct

File: src/System.Management.Automation/engine/EventManager.cs [hunk 1]
```diff
@@ -2042,6 +2042,19 @@ private ScriptBlock CreateBoundScriptBlock(ScriptBlock scriptAction)
...
+        /// Determines whether the specified object is equal to the current object.
+        /// </summary>
+        /// <param name="obj">The object to compare with the current object.</param>
+        /// <returns>
+        /// <see langword="true"/> if the specified object is equal to the current object;
+        /// otherwise, <see langword="false"/>.
+        /// </returns>
+        public override bool Equals(object obj)
+        {
+            return Equals(obj as PSEventSubscriber);
```
Comment (vexx32) [inline src/System.Management.Automation/engine/EventManager.cs:2055]: From the doc you linked the suggested form was something like this instead. Is there a significant difference either way? 🤔 

```suggestion
            return obj is PSEventSubscriber es && Equals(es);
```

Comment (xtqqczze) [reply src/System.Management.Automation/engine/EventManager.cs:2055]: Currently the method replies on the null check in the `Equals(PSEventSubscriber)` method immediately below for correctness, I guess your suggested change makes sense though.

https://github.com/PowerShell/PowerShell/blob/72b8c61c7d74c9631b80a05223d4bc01852ab884/src/System.Management.Automation/engine/EventManager.cs#L2064-L2072

**Comment by iSazonov** [comment]
@xtqqczze Please add some tests as Committee ask.

**Comment by xtqqczze** [comment]
@iSazonov I do not know even where to begin with adding these tests :(

**Comment by iSazonov** [comment]
@xtqqczze xUnit tests are in test\xUnit\csharp folder.

**Review by TravisEz13 — APPROVED** [review]

**Pull request merged**